### DISCUSSION TITLE
How to find, (re)use and create ontology design patterns

### DISCUSSION THREAD

**Discussion by StroemPhi** [jamesaoverton/obook#199] [Tutorial requests]
As an ontology developer, I want to reuse as much design patterns as possible in order to be OBO compliant and FAIR. For this I need to know where previous design patterns are documented and how they are to be implemented (e.g. ODK, DOSDP & ROBOT). I also would love to have a central place where I can read up on and/or ask my questions about these desing patterns, as well as where to make new pattern requests. What I'm hoping for is a single point of truth or registry, if you will, for OBO standard patterns. IDK if this is possible, but having these patterns there already also present as assets (OWL, OFN, ...) might also be helpful for the standardization of core import modules. I know that there is http://ontologydesignpatterns.org, but what I mean is more high level and context dependent. 

An example:
I need to build an (application) ontology that is to be used in a research data management context for the proper annotation of various kinds of data files associated with a certain kind of assay, e.g. NMR spectroscopy. For this, I can already reuse lots of terms from OBI and IAO (assay, study, investigation, protocol, plan specification, ...). Building the import modules needed in way that is lean but includes all necessary axioms, is very challanging. I think having a "pattern registry" or FAQ could be  great help in this regard.

**Comment by StroemPhi** [comment]
With regard to the implementation of design patterns via templates there is already this: https://oboacademy.github.io/obook/lesson/templates-for-obo/. 
However, what I have in mind is more of a documentation source that bundles things like this: http://obi-ontology.org/docs/data-intro/

**Comment by matentzn** [comment]
What you are asking has been our dream for a while, but to be honest, there is not enough drive to realise this in any short or midterm timeline. The current way is this:

1. The ontology that hosts the terms which are the "SUBJECTS" of the axioms (so the A in `A subclassOf R some C`) is considered responsible for managing the patterns. This works ok for OBI, CL, Uberon, RO (to some extend), uPheno and maybe a handful others.
2. New pattern request and questions go to their respective trackers
3. A centralised place is at the moment out of reach.
4. How can OBOOK help here?

**Comment by StroemPhi** [comment]
Yeah, I see the hard work behind this unfortunately. And I guess I'm asking for too much. But when I say I'd love to have a single point of truth, I don't mean this to be the place where edits are made, but more of a hub mirroring the relevant docs. So 1. and 2. should off course be the place where patterns and their documentation are worked on. 

Regarding 3. & 4.:
How about tagging all the issues in the respective trackers that are about `pattern definition` or `pattern clarification/discussion` with an appropriate keyword, so that an automatic aggregation over these repos in the OBOOK can take place and a keen learner would thus get faster to the desired links? Or, maybe have ODK generated templates in the `doc` folder that help the curators/developers add the pattern documentation more easily? In how far can automatic graph visualisations be leveraged for this, where one can set the core nodes to be visualized for the pattern in question, so that devs can produce such helpful images quick and easy, as in the OBI example linked to above?

If we are talking about a dream here anyway, maybe verbalizing my version through such questions might help in the long term. If this is the wrong place to ask these kind of questions, please let me know and I appologize in advance. I really appreciate the efforts behind you OBOOK creators and just thought this might be helpful. 😃

**Reply by matentzn** [reply]
You know what I would kinda find even better as a first pass? A simple page: "Design Patterns used across OBO" linking, for now, to all design patterns we know of (CL, Uberon, GO, OBI, OBA, TO, to name a few). This page can also serve as a first pass intro into OBO design pattern thinking. Later, we can try to generate graphics from the collected patterns that illustrate how classes should relate to each other, but for now, I think, just a collection really is what you want.

**Reply by matentzn** [reply]
If you can make a PR with a draft page on OBOOK, I will flesh it out!

**Reply by matentzn** [reply]
I toyed once with this idea (a standalone documentation page for all OBO patterns), but there is not enough volunteering power to get this up and running: https://github.com/OBOFoundry/obo-patterns.github.io

**Reply by StroemPhi** [reply]
Just to make sure, with the PR you mean your proposal to just have a link collection page to the DOSDP or ROBOT pattern templates of all OBO ontology repos?

**Reply by StroemPhi** [reply]
And should this be going into the direction of a reference, explanation, how-to, tutorial or lessen? Just asking to know where to put the .md.

**Reply by matentzn** [reply]
Yes, just a markdown page in the "reference" section of obook for now - we will move it out later. 

Get inspired in the "usage" section here: https://docs.google.com/document/d/1KXv-Tm3HvypjucFZeLLPjCdgkof5f0Re5_5GbFfQLWU/edit#heading=h.533dqok37h0
